File: packages/python/navaia_forge/resources/tasks.py
"""Task resource."""

from __future__ import annotations

import time
from typing import Any

from ..errors import TimeoutError as NfTimeoutError
from ..types import Task, TaskStatus
from ._base import ResourceBase, parse_list, parse_model

_TERMINAL_STATES: frozenset[str] = frozenset({"done", "failed", "rejected"})


class TasksResource(ResourceBase):
    """CRUD + helper operations for tasks."""

    def list(
        self, workforce_id: str, *, status: TaskStatus | str | None = None
    ) -> list[Task]:
        """List tasks for a workforce, optionally filtered by status."""
        params: dict[str, Any] = {}
        if status is not None:
            params["status"] = status
        return parse_list(
            Task,
            self._http.get_list(
                f"/workforces/{workforce_id}/tasks",
                params=params or None,
            ),
        )

    def get(self, task_id: str) -> Task:
        """Fetch a single task by id."""
        return parse_model(Task, self._http.get(f"/tasks/{task_id}"))

    def create(
        self,
        workforce_id: str,
        title: str,
        *,
        description: str = "",
        agent_id: str | None = None,
        priority: str = "standard",
        metadata: dict[str, Any] | None = None,
    ) -> Task:
        """Create a task."""
        body: dict[str, Any] = {
            "workforce_id": workforce_id,
            "title": title,
            "description": description,
            "priority": priority,
        }
        if agent_id is not None:
            body["agent_id"] = agent_id
        if metadata is not None:
            body["metadata_json"] = metadata
        return parse_model(
            Task, self._http.post(f"/workforces/{workforce_id}/tasks", body)
        )

    def approve(self, task_id: str) -> Task:
        """Approve a task that is waiting on human approval."""
        return parse_model(Task, self._http.post(f"/tasks/{task_id}/approve"))

    def reject(self, task_id: str, reason: str = "") -> Task:
        """Reject a task awaiting approval, with optional reason."""
        return parse_model(
            Task, self._http.post(f"/tasks/{task_id}/reject", {"reason": reason})
        )

    def retry(self, task_id: str) -> Task:
        """Retry a failed task."""
        return parse_model(Task, self._http.post(f"/tasks/{task_id}/retry"))

    def wait_for_completion(
        self,
        task_id: str,
        *,
        poll_interval: float = 5.0,
        timeout: float = 300.0,
    ) -> Task:
        """Poll until the task reaches a terminal state.

        Raises:
            TimeoutError: If the task does not reach a terminal state in time.
        """
        start = time.monotonic()
        while time.monotonic() - start < timeout:
            task = self.get(task_id)
            if task.status in _TERMINAL_STATES:
                return task
            time.sleep(poll_interval)
        raise NfTimeoutError(f"Task {task_id} did not complete within {timeout}s")

File: packages/javascript/src/resources/tasks.ts
/**
 * Task resource — CRUD, approval/rejection, retry, and polling.
 */

import { TimeoutError } from "../errors.js";
import { get, getList, post } from "../http.js";
import type {
  ResolvedConfig,
  Task,
  TaskCreate,
  TaskLog,
  WaitForCompletionOptions,
} from "../types.js";

/** Terminal task statuses that stop polling. */
const TERMINAL_STATUSES: ReadonlySet<string> = new Set([
  "done",
  "failed",
  "cancelled",
  "waiting_question",
  "waiting_plan",
  "waiting_blocked",
]);

export class TaskResource {
  private readonly config: ResolvedConfig;

  constructor(config: ResolvedConfig) {
    this.config = config;
  }

  /**
   * List tasks for a workforce, optionally filtered by status.
   *
   * @param workforceId - The workforce whose tasks to list.
   * @param status - Optional status filter.
   */
  list(workforceId: string, status?: string): Promise<Task[]> {
    const params: Record<string, string> = {};
    if (status) {
      params["status"] = status;
    }
    return getList<Task>(
      this.config,
      `/workforces/${workforceId}/tasks`,
      params,
    );
  }

  /** Get a single task by ID. */
  get(taskId: string): Promise<Task> {
    return get<Task>(this.config, `/tasks/${taskId}`);
  }

  /** Create a new task.
   *
   * Posts to ``/workforces/{workforce_id}/tasks`` (matching the live backend
   * route — ``POST /tasks`` returns 404 in production).
   */
  create(data: TaskCreate): Promise<Task> {
    return post<Task>(
      this.config,
      `/workforces/${data.workforce_id}/tasks`,
      data,
    );
  }

  /** Approve a task that is waiting for plan/question approval. */
  approve(taskId: string): Promise<Task> {
    return post<Task>(this.config, `/tasks/${taskId}/approve`);
  }

  /**
   * Reject a task or plan.
   *
   * @param taskId - The task to reject.
   * @param reason - Optional rejection reason.
   */
  reject(taskId: string, reason = ""): Promise<Task> {
    return post<Task>(this.config, `/tasks/${taskId}/reject`, { reason });
  }

  /** Retry a failed task. */
  retry(taskId: string): Promise<Task> {
    return post<Task>(this.config, `/tasks/${taskId}/retry`);
  }

  /** Get the event log for a task. */
  logs(taskId: string): Promise<TaskLog[]> {
    return get<TaskLog[]>(this.config, `/tasks/${taskId}/logs`);
  }

  /**
   * Poll until a task reaches a terminal state.
   *
   * @param taskId - The task to wait for.
   * @param options - Polling interval and timeout configuration.
   * @returns The completed task.
   * @throws {@link TimeoutError} if the task does not complete within the timeout.
   *
   * @example
   * ```ts
   * const result = await nf.tasks.waitForCompletion(task.id, {
   *   pollInterval: 3000,
   *   timeout: 120_000,
   * });
   * ```
   */
  async waitForCompletion(
    taskId: string,
    options: WaitForCompletionOptions = {},
  ): Promise<Task> {
    const pollInterval = options.pollInterval ?? 5_000;
    const timeout = options.timeout ?? 300_000;
    const start = Date.now();

    while (Date.now() - start < timeout) {
      const task = await this.get(taskId);
      if (TERMINAL_STATUSES.has(task.status)) {
        return task;
      }
      await sleep(pollInterval);
    }

    throw new TimeoutError(
      `Task ${taskId} did not complete within ${timeout}ms`,
    );
  }
}

/** Promise-based sleep helper. */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
